### PR TITLE
AO3-6381 Restrict admin access to spam page

### PR DESCRIPTION
**Pull request by Cesium-Ice**
# Pull Request Checklist

* [x] Have you read ["How to write the perfect pull request"](https://github.blog/2015-01-21-how-to-write-the-perfect-pull-request/)?
* [x] Have you read the [contributing guidelines](https://github.com/otwcode/otwarchive/blob/master/CONTRIBUTING.md)?
* [x] Have you added [tests for any changed functionality](https://github.com/otwcode/otwarchive/wiki/Automated-Testing)?
* [x] Have you added the [Jira](https://otwarchive.atlassian.net) issue number
  as the *first* thing in your pull request title (e.g. `AO3-1234 Fix thing`)
* [x] Do you have fewer than 5 pull requests already open? If not, please wait
  until they are reviewed and merged before creating new pull requests.

## Issue

https://otwarchive.atlassian.net/browse/AO3-6381

## Purpose

What does this PR do?
Restrict admin access to spam

## Testing Instructions

How can the Archive's QA team verify that this is working as you intended?
Description 
Log in as an admin with the policy_and_abuse or superadmin role

You should see Spam in the admin navigation and it should take you to https://test.archiveofourown.org/admin/spam

You should be able to perform the various marking/unmarking as spam activities on this page as well as the Confirmed Spam and Approved Works pages

 

Log in as an admin without the policy_and_abuse or superadmin role

You should not see Spam in the admin navigation and should not be able to access https://test.archiveofourown.org/admin/spam 

If you have a Jira account with access, please update or comment on the issue
with any new or missing testing instructions instead.

## References

Are there other relevant issues/pull requests/mailing list discussions?

## Credit

What name and pronouns should we use to credit you in the [Archive of Our Own's Release Notes](https://archiveofourown.org/admin_posts?tag=1)?
Cesium-Ice,  she/her
If you have a Jira account, please include the same name in the "Full name"
field on your Jira profile, so we can assign you the issues you're working on.

*Please note that if you do not fill in this section, we will use your GitHub account name and
they/them pronouns.*

### CODE REVIEW

**Comment by Cesium-Ice** [comment]
The failed tests involve gifts and orphaning, which I did not make any changes to, so I am unsure why they are failing.

**Comment by sarken** [comment]
We have some [tests that fail intermittently](https://github.com/otwcode/otwarchive/wiki/Automated-Testing#Known_Test_Failures), so if you see a failure that's unrelated to your changes, there's a good chance it's not your fault. Someone will notice the red "x" and rerun the failed tests for you to make sure, though!

File: app/views/admin/_header.html.erb [hunk 1]
```diff
@@ -33,11 +33,15 @@
...
-    <li><%= link_to ts("Settings", key: "header"), admin_settings_path %></li>
+ 
+  <% if policy(:spam).index? %>
+    <li><%= link_to ts("Spam", key: "header"), admin_spam_index_path %></li>
   <% end %>
 
+  <% #if policy(AdminSetting).can_view_settings? %>
+    <li><%= link_to ts("Settings", key: "header"), admin_settings_path %></li>
+  <% #end %>
+
```
Comment (sarken) [inline app/views/admin/_header.html.erb:44]: I think these changes were unintentional.

File: app/views/admin/spam/index.html.erb [hunk 2]
```diff
@@ -4,15 +4,17 @@
...
+    <li>
+      <%= span_if_current ts("Unreviewed Works"), admin_spam_index_path, params[:show].blank? %>
+    </li>
+    <li>
+      <%= span_if_current ts("Confirmed Spam"), { show: "reviewed" } %>
+    </li>
+    <li>
+      <%= span_if_current ts("Approved Works"), { show: "approved" } %>
+    </li>
+  <% end %>
```
Comment (sarken) [inline app/views/admin/spam/index.html.erb:17]: I don't think you need this. Everyone who can access the index should be able to access the other pages as well.

File: spec/controllers/admin/spam_controller_spec.rb [hunk 3]
```diff
@@ -48,28 +54,36 @@
...
-    context "when logged in as admin" do
-      it "marks moderated workd as reviewed, marks works as spam, hides the works, and redirects with notice" do
-        FactoryBot.create_list(:moderated_work, 3)
-        fake_login_admin(admin)
-        post :bulk_update, params: { spam: ModeratedWork.all.map(&:id) }
+    %w[policy_and_abuse superadmin].each do |role|
+      context "when logged in as an admin with #{role} role" do
+        let(:admin) { create(:admin, roles: [role]) }
+
+        it "marks moderated workd as reviewed, marks works as spam, hides the works, and redirects with notice" do
```
Comment (sarken) [inline spec/controllers/admin/spam_controller_spec.rb:75]: Could you fix the existing typo in this line? `workd` should be `works`.

File: app/policies/spam_policy.rb [hunk 4]
```diff
@@ -0,0 +1,9 @@
+class SpamPolicy < ApplicationPolicy
```
Comment (sarken) [inline app/policies/spam_policy.rb:1]: I know these are called the spam pages, but maybe this should be the `ModeratedWorkPolicy` instead, so it's more obvious it's not related to spam comments. Or... should we use the `UserCreationPolicy` for this, considering it already controls the ability to mark works as spam? 

I think I'm going to solicit some additional opinions on this.

**Review by sarken — CHANGES_REQUESTED** [review]

**Comment by Cesium-Ice** [comment]
I accidentally made a commit to this pull request with the wrong account, is there a way to revert that or should I close this request and make a new one?

**Comment by sarken** [comment]
You can try [reverting the merge commit](https://stackoverflow.com/a/7100005) , which I think is the erroneous one (it's hard to tell on mobile), but it might be easier to close and start fresh.

File: app/views/admin/_header.html.erb (hunk 1 above)
Comment (Cesium-Ice) [reply app/views/admin/_header.html.erb:44]: yup

**Comment by Cesium-Ice** [comment]
Closing to open another PR

**Comment by Cesium-Ice** [comment]
New PR: https://github.com/otwcode/otwarchive/pull/4340

**Pull request closed without merging**